core-patch/src/main/scala/eu/ostrzyciel/jelly/core/patch/internal/PatchDecoderImpl.scala

```diff
@@ -0,0 +1,188 @@
+    converter: ProtoDecoderConverter[TNode, TDatatype, ?, ?],
+    handler: TriplePatchHandler[TNode],
+    supportedOptions: RdfPatchOptions,
+  ) extends PatchDecoderImpl[TNode, TDatatype](converter, handler, supportedOptions):
+
+    override protected def handleOptions(opt: RdfPatchOptions): Unit =
+      if !opt.statementType.isTriples then
+        throw new RdfProtoDeserializationError(
+          f"Incoming stream with statement type ${opt.statementType} cannot be decoded by this " +
+          f"decoder. Only TRIPLES streams are accepted.")
+      super.handleOptions(opt)
+
+    override protected def handleStatementAdd(statement: RdfQuad): Unit =
+      handler.addTriple(
+        convertTermWrapped(statement.subject, lastSubject),
+        convertTermWrapped(statement.predicate, lastPredicate),
+        convertTermWrapped(statement.`object`, lastObject),
+      )
+
+    override protected def handleStatementDelete(statement: RdfQuad): Unit =
+      handler.deleteTriple(
+        convertTermWrapped(statement.subject, lastSubject),
+        convertTermWrapped(statement.predicate, lastPredicate),
+        convertTermWrapped(statement.`object`, lastObject),
+      )
+
+  private[core] final class QuadsDecoder[TNode, TDatatype : ClassTag](
+    converter: ProtoDecoderConverter[TNode, TDatatype, ?, ?],
+    handler: QuadPatchHandler[TNode],
+    supportedOptions: RdfPatchOptions,
+  ) extends PatchDecoderImpl[TNode, TDatatype](converter, handler, supportedOptions):
+
+    override protected def handleOptions(opt: RdfPatchOptions): Unit =
+      if !opt.statementType.isQuads then
+        throw new RdfProtoDeserializationError(
+          f"Incoming stream with statement type ${opt.statementType} cannot be decoded by this " +
+          f"decoder. Only QUADS streams are accepted.")
+      super.handleOptions(opt)
+
+    override protected def handleStatementAdd(statement: RdfQuad): Unit =
+      handler.addQuad(
+        convertTermWrapped(statement.subject, lastSubject),
+        convertTermWrapped(statement.predicate, lastPredicate),
+        convertTermWrapped(statement.`object`, lastObject),
+        convertGraphTermWrapped(statement.graph),
+      )
+
+    override protected def handleStatementDelete(statement: RdfQuad): Unit =
+      handler.deleteQuad(
+        convertTermWrapped(statement.subject, lastSubject),
+        convertTermWrapped(statement.predicate, lastPredicate),
+        convertTermWrapped(statement.`object`, lastObject),
+        convertGraphTermWrapped(statement.graph),
+      )
+
+  private[core] final class AnyStatementDecoder[TNode, TDatatype : ClassTag](
+    converter: ProtoDecoderConverter[TNode, TDatatype, ?, ?],
+    handler: AnyPatchHandler[TNode],
+    supportedOptions: RdfPatchOptions,
+  ) extends PatchDecoderImpl[TNode, TDatatype](converter, handler, supportedOptions):
+    private var inner: PatchDecoderImpl[TNode, TDatatype] = uninitialized
```

Comment:
not sure why this class needs an inner instance of PatchDecoderImpl when it already extends the same class and it is not a pattern I see in other decoders

Reply:
I guess I could simplify that. I'll see what I can do here.